astropy/nddata/mixins/ndslicing.py

```diff
@@ -11,29 +18,224 @@ class NDSlicingMixin(object):
+        array([False,  True], dtype=bool)
+        >>> # But be aware that the slicing only returns references not copies
+        >>> nd3 = nd2[1:3]
+        >>> nd3.data[0] = 100
+        >>> nd2
+        NDDataSliceable([  1, 100,   3,   4,   5])
+        >>> # If the mask is not a numpy array a warning is issued
+        >>> nd4 = NDDataSliceable(nd, mask=[False, True])
+        >>> nd5 = nd4[0:2] # doctest: +SKIP
+        INFO: Mask is considered not sliceable. [astropy.nddata.mixins.ndslicing]
+        INFO: Therefore the mask will not be sliced. [astropy.nddata.mixins.ndslicing]
+        >>> # but the data is sliced
+        >>> nd5 # doctest: +SKIP
+        NDDataSliceable([  1, 100])
+        >>> # and the mask is just the mask without slicing
+        >>> nd5.mask # doctest: +SKIP
+        [False, True]
+
     """
     def __getitem__(self, item):
-
+        # Abort slicing if the data is scalar
+        if self._data.shape == ():
+            raise TypeError('Scalars cannot be sliced.')
+        # Slice the data but everything else slice in self._slice.
         new_data = self.data[item]
+        kwargs = self._slice(item)
+        return self.__class__(new_data, **kwargs)
+
+    def _slice(self, item):
+        """
+        Collects the sliced attributes and passes them back as ``kwarg`` for
+        the return after slicing.
+
+        It passes uncertainty, mask and wcs to their appropriate ``_slice_*``
+        method, while `meta` and `unit` are simply taken from the original. The
+        data is assumed to be sliceable and is sliced already before.
+
+        Where possible the return is *not* a copy of the data.
+
+        Parameters
+        ----------
+        item: Slice
+            The slice passed to ``__getitem__``.
 
-        if self.uncertainty is not None:
-            new_uncertainty = self.uncertainty[item]
+        Returns
+        -------
+        kwargs: `dict`
+            Containing all the attributes except data after slicing - ready to
+            feed them into the ``init``.
+        """
+        kwargs = {}
+        # Try to slice some attributes
+        kwargs['uncertainty'] = self._slice_uncertainty(item)
+        kwargs['mask'] = self._slice_mask(item)
+        kwargs['wcs'] = self._slice_wcs(item)
+        # Attributes which are copied and not intended to be sliced
+        kwargs['unit'] = self._unit
+        kwargs['meta'] = self._meta
+        return kwargs
+
+    def _slice_uncertainty(self, item):
+        # Uncertainty is not set
+        if self._uncertainty is None:
+            return None
+        elif isinstance(self._uncertainty, np.ndarray):
+            if self._uncertainty.shape == self.data.shape:
+                return self._uncertainty[item]
+            else:
+                log.info("Uncertainty has not the same shape as data.")
+        elif isinstance(self._uncertainty, NDUncertainty):
+            if self._uncertainty.array.shape == self.data.shape:
```

Comment:
Currently `NDUncertainty` defined no setter for the `array` (which contains the uncertainty values) so subclasses might not implement shape checks in the setter or `init`. So there were two possibilities:
1.) enforcing the uncertainty to have the same shape as the data, which is a tremendeous effort because there might be use-cases were we only need the uncertainty to be broadcastable to the data's shape. 
2.) which brings me to the second point: If we don't enforce the shapes to match we would need to implement how slicing affects something that is only broadcastable to the data shape, this is not so complicated but the most common case is, when the uncertainty is either a scalar or a ndarray with the same shape as the data. In either case the current slicing works (ok it issues a warning if it is a scalar but it does not try to slice it).

If someone really want's to slice more complicated uncertainties (which are only broadcastable to data shape or as you pointed out with the mask some lazy-evaluation function) the subclass has to override the `_slice_uncertainty` method.

Or is the requirement that the uncertainty has the same shape as the data something that should be enforced?